servant/src/Servant/API/UVerb/OpenUnion.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ExplicitNamespaces #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | Type-level code for implementing and using 'UVerb'.
+module Servant.API.UVerb.OpenUnion
+( IsMember
+, Unique
+, inject
+)
+where
+
+import Data.SOP.Constraint
+import Data.SOP.NS
+import Data.Type.Bool (If)
+import Data.Type.Equality (type (==))
+import GHC.TypeLits
+
+
+-- * Stuff stolen from 'Data.WorldPeace" but for generics-sop
+-- TODO: could much of this go into sop-core?
+
+type IsMember (a :: u) (as :: [u]) = (CheckElemIsMember a as, UElem a as (RIndex a as))
+
+type family Contains (as :: [k]) (bs :: [k]) :: Constraint where
+  Contains '[] _ = ()
+  Contains (a ': as) bs = (IsMember a bs, Contains as bs)
+
+data Nat_ = S_ Nat_ | Z_
+
+
+-- | TODO: we probably can make do without this.
```

Comment:
This TODO should be resolved.  the `Index` part is never used anywhere.  We can get away with:

```haskell
class Elem x xs where
  inject :: f x -> NS f xs
  eject :: NS f xs -> Maybe (f x)

-- base case. we found `x` in `xs`
instance Elem x (x ': xs) where
  inject x = Z x
  eject = \case
    Z x -> Just x
    _ -> Nothing

-- recursive case: We did not find `x` but it might be in `xs`
instance Elem x xs => Elem x (y ': xs) where
  inject = S . inject
  eject = \case
    Z _ -> Nothing
    S z -> eject z
```


This is also now generic enough that we can probably upstream it to `sop-core` as the typeclass:

```
class NSElem (x :: k) (xs :: [k]) where
  injection :: Injection f xs x
  ejection ::  Ejection f xs x
```

@kosmikus  what do you think?  Is such a typeclass useful in `sop-core` ? it seems a bit weird to keep it in `servant` but it's ok for now

Reply:
the pair `inject,eject` form a `Prism`   even . Not sure what to do with that observation :)

Reply:
Ah it actually is a prism in world-peace :) http://hackage.haskell.org/package/world-peace-1.0.2.0/docs/src/Data.WorldPeace.Union.html#UElem

Reply:
From their source code there is a note:

```
-- This is safe because of the 'RIndex' contraint. This 'RIndex' constraint
-- tells us that there /actually is/ an @a@ in @as@ at index @i@.
```
so perhaps I am indeed missing something and it _is_ needed. Type level programming is confusing

Reply:
Without the `RIndex` this complains about overlapping instances. so perhaps it _is_ needed. Let me give it a bit more thought tonight :)

Reply:
Note: I deleted some comments above to avoid confusion. The above code I'll add to the PR to simplify things tomorrow morning :)

Reply:
@maksbotan what are your thoughts on this? I wrote this quite a while ago; and not sure anymore if it's equivalent with what we have.  I can have a look myself later; but if somebody else could have a look that'd be great

Reply:
@arianvp looks like your idea works. See here: https://github.com/zerobuzz/servant/compare/uverb...maksbotan:uverb

(feel free to cherry-pick those commits, I've also fixed compilation of the cookbook)